flink-filesystems/flink-azure-fs-hadoop/src/main/java/org/apache/flink/fs/azurefs/AzureBlobFsRecoverableDataOutputStream.java

```diff
@@ -0,0 +1,299 @@
+
+    /**
+     * Implementation of a committer for the Hadoop File System abstraction. This implementation
+     * commits by renaming the temp file to the final file path. The temp file is truncated before
+     * renaming in case there is trailing garbage data.
+     */
+    static class ABFSCommitter implements Committer {
+
+        private final FileSystem fs;
+        private final HadoopFsRecoverable recoverable;
+
+        ABFSCommitter(FileSystem fs, HadoopFsRecoverable recoverable) {
+            this.fs = checkNotNull(fs);
+            this.recoverable = checkNotNull(recoverable);
+        }
+
+        @Override
+        public void commit() throws IOException {
+            final Path src = recoverable.tempFile();
+            final Path dest = recoverable.targetFile();
+            final long expectedLength = recoverable.offset();
+            FileStatus srcStatus = null;
+            try {
+                srcStatus = fs.getFileStatus(src);
+            } catch (FileNotFoundException fnfe) {
+                // srcStatus will be null
+            } catch (IOException e) {
+                throw new IOException("Cannot clean commit: Staging file does not exist.");
+            }
+            if (srcStatus != null) {
+                LOG.debug(
+                        "The srcStatus is {} and exp length is {}",
+                        srcStatus.getLen(),
+                        expectedLength);
+                if (srcStatus.getLen() != expectedLength) {
+                    LOG.error(
+                            "The src file {} with length {} does not match the expected length {}",
+                            src,
+                            srcStatus.getLen(),
+                            expectedLength);
+                    throw new IOException(
+                            "The src file "
+                                    + src
+                                    + " with length "
+                                    + srcStatus.getLen()
+                                    + " "
+                                    + "does not match the expected length "
+                                    + expectedLength);
+                }
+                try {
+                    fs.rename(src, dest);
+                } catch (IOException e) {
+                    throw new IOException(
+                            "Committing file by rename failed: " + src + " to " + dest, e);
+                }
+            } else if (!fs.exists(dest)) {
+                // neither exists - that can be a sign of
+                //   - (1) a serious problem (file system loss of data)
+                //   - (2) a recovery of a savepoint that is some time old and the users
+                //         removed the files in the meantime.
+
+                // TODO how to handle this?
+                // We probably need an option for users whether this should log,
+                // or result in an exception or unrecoverable exception
```

Comment:
I think we should explicitly throw an exception here, because the data cannot be committed.